draft-janfred-radext-radiusdtls-bis.md

```diff
@@ -183,26 +183,24 @@ While the client may be able to deduce the operational state of the local server
 Within RADIUS, proxies typically only forward traffic between the NAS and RADIUS servers, and they do not generate their own response.
 As a result, when a NAS does not receive a response to a request, this could be the result of packet loss between the NAS and proxy, a problem on the proxy, loss between the RADIUS proxy and server, or a problem with the server.
 
-When UDP is used as a transport protocol, the absence of a reply can cause a client to deduce (incorrectly) that the proxy is unavailable.
+The absence of a reply can cause a client to deduce (incorrectly) that the proxy is unavailable.
 The client could then fail over to another server or conclude that no "live" servers are available (OKAY state in {{!RFC3539}}, Appendix A).
 This situation is made even worse when requests are sent through a proxy to multiple destinations.
 Failures in one destination may result in service outages for other destinations, if the client erroneously believes that the proxy is unresponsive.
 
-For RADIUS/TLS, it is RECOMMENDED that implementations utilize the existence of a TCP connection along with the application-layer watchdog defined in {{RFC3539}}, Section 3.4 to determine that the server is "live".
-RADIUS/TLS clients MUST mark a connection DOWN, if the network stack indicates that the connection is no longer active.
-If the network stack indicates that the connection is still active, clients MUST NOT decide that it is down until the application-layer watchdog algorithm has marked it DOWN.
-RADIUS/TLS clients MUST NOT decide that a RADIUS/TLS server is unresponsive until all TLS connections to it have been marked down.[^contradiction]{:jf}
+It is REQUIRED that implementations utilize the existence of a TCP/DTLS connection along with the application-layer watchdog defined in {{RFC3539}}, Section 3.4 to determine that the server liveliness.
```

Comment:
If you receive an error on `write()` or via an event in your event loop, the connection is DOWN, i.e. the underlying connection is gone, and you'll need to re-establish the connection.  There's no point in waiting for an application level watchdog to indicate that the connection is down, attempting to write to a closed connection will just result in further errors.

I dislike the text about waiting for all connections to fail.  That's an implementors decision.  If 9/10 connections have been closed, and new connections are failing, you can be fairly sure that the server is unavailable, you don't need to wait for the last connection to time out.

Reply:
The wording in the text can be read like you do; but I think the intention was different (and thus it should be rewritten for clarity.

If you got a positive indication from the network stack that the TCP session is gone (e.g. after receiving an RST) then there is no point in waiting for anything *on this particular connection*. It should be marked as down and re-established immediately following the event.

The more interesting case is when TCP is not yet down, just doesn't get any application layer responses. In that case, waiting for the application-layer is the thing to do (for any one connection in isolation).

No guessing should be done about "9/10 are not responding, so the 10th is certainly also down". There can be situations in proxy environments where a proxy is waiting for downstream RADIUS servers with a long response timeout; if 9 of 10 of its downstreams take their time, then 9 of 10 TCP sessions from the client to that proxy hang; but it is still possible that a tenth connection to the same proxy but another downstream is healthy and replying.

I guess in a pseudocode we would write this with a short-circuit OR:

```
foreach (TCP socket in use) {
  if (socket reported closed by OS || application-layer watchdog fails) {
     mark connection failed;
     establish new conn;
  } else {
   wait;
  }
} 
```